Match on Go interface type does not emit a type switch

Given a Lisette struct that implements the bubbletea `tea.Model` interface:

```rs
import "go:charm.land/bubbletea/v2"
import "go:github.com/charmbracelet/ultraviolet"

struct Model { width: int, height: int }

struct NextWordMsg {}

impl Model {
  fn Init(self: Model) -> Option<tea.Cmd> { None }

  fn Update(self: Model, msg: uv.Event) -> (tea.Model, Option<tea.Cmd>) {
    match msg {
      tea.WindowSizeMsg { Width: width, Height: height } => (
        Model { width, height, ..self },
        None,
      ),
      tea.KeyPressMsg { .. } => (Model { ..self }, None),
      NextWordMsg {} => (Model { ..self }, None),
      _ => (Model { ..self }, None),
    }
  }

  fn View(self: Model) -> tea.View { tea.NewView("") }
}

fn main() {
  if let Err(err) = tea.NewProgram(Model { width: 0, height: 0 }).Run() { panic(err) }
}
```

The type switch is missing:

```go
func (m Model) Update(msg uv.Event) lisette.Tuple2[tea.Model, lisette.Option[tea.Cmd]] {
    width := msg.Width
    height := msg.Height
    copy_1 := m
    copy_1.width = width
    copy_1.height = height
    return lisette.MakeTuple2[tea.Model, lisette.Option[tea.Cmd]](
        _lisAdapter_Model_Model_0{inner: copy_1},
        lisette.MakeOptionNone[tea.Cmd](),
    )
}
```

Leading to:

```
./main.go:77:15: msg.Width undefined (type uv.Event has no field or method Width)
./main.go:78:16: msg.Height undefined (type uv.Event has no field or method Height)
```

Version:

```
lisette 0.1.14 (go 1.25.5)
```

--- a/crates/emit/src/go/patterns/decision_tree.rs
+++ b/crates/emit/src/go/patterns/decision_tree.rs
@@ -88,6 +88,9 @@ pub(crate) enum Check {
     /// Each inner `Vec<Check>` is one alternative (checks ANDed together);
     /// the alternatives are ORed.
     Or { alternatives: Vec<Vec<Check>> },
+    /// Go interface type assertion: the value at `path` implements `go_type`.
+    /// Emitted as a case label in a `switch x := x.(type)` statement.
+    TypeAssert { path: AccessPath, go_type: String },
 }
 
 impl Check {
@@ -143,6 +146,9 @@ impl Check {
                     joined
                 }
             }
+            // Type assertions are emitted via type switches, not boolean conditions.
+            // This path is only reached when a guard prevents switch compilation.
+            Check::TypeAssert { .. } => "false".to_string(),
         }
     }
 
@@ -152,7 +158,8 @@ impl Check {
             Check::EnumTag { path, .. }
             | Check::Literal { path, .. }
             | Check::SliceLenEq { path, .. }
-            | Check::SliceLenGe { path, .. } => Some(path),
+            | Check::SliceLenGe { path, .. }
+            | Check::TypeAssert { path, .. } => Some(path),
             Check::Or { .. } => None,
         }
     }
@@ -173,6 +180,13 @@ impl Check {
     pub(crate) fn as_literal(&self) -> Option<&str> {
         match self {
             Check::Literal { go_literal, .. } => Some(go_literal),
+            _ => None,
+        }
+    }
+
+    pub(crate) fn as_type_assert(&self) -> Option<&str> {
+        match self {
+            Check::TypeAssert { go_type, .. } => Some(go_type),
             _ => None,
         }
     }
@@ -246,6 +260,8 @@ pub(crate) enum SwitchKind {
     EnumTag,
     /// Switch on value directly — literals, booleans, units
     Value,
+    /// Switch on dynamic Go type — `switch x := x.(type)`
+    TypeSwitch,
 }
 
 /// A single branch in a Switch node.
@@ -320,6 +336,8 @@ fn try_build_switch(arms: &[ArmInfo]) -> Option<Decision> {
         (SwitchKind::EnumTag, first_check.path()?.clone())
     } else if first_check.as_literal().is_some() {
         (SwitchKind::Value, first_check.path()?.clone())
+    } else if first_check.as_type_assert().is_some() {
+        (SwitchKind::TypeSwitch, first_check.path()?.clone())
     } else {
         return None;
     };
@@ -328,7 +346,8 @@ fn try_build_switch(arms: &[ArmInfo]) -> Option<Decision> {
         if arm.checks.is_empty() {
             continue;
         }
-        if arm.has_guard {
+        // Type switches handle guards inside the case body; other switches cannot.
+        if arm.has_guard && !matches!(kind, SwitchKind::TypeSwitch) {
             return None;
         }
         let first = arm.checks.first()?;
@@ -336,6 +355,7 @@ fn try_build_switch(arms: &[ArmInfo]) -> Option<Decision> {
         let is_compatible = match &kind {
             SwitchKind::EnumTag => first.as_enum_tag().is_some(),
             SwitchKind::Value => first.as_literal().is_some() && arm.checks.len() == 1,
+            SwitchKind::TypeSwitch => first.as_type_assert().is_some(),
         };
         if !is_compatible {
             return None;
@@ -365,6 +385,10 @@ fn try_build_switch(arms: &[ArmInfo]) -> Option<Decision> {
                 let lit = first_check.as_literal().unwrap();
                 (lit.to_string(), false)
             }
+            SwitchKind::TypeSwitch => {
+                let go_type = first_check.as_type_assert().unwrap();
+                (go_type.to_string(), false)
+            }
         };
 
         let inner_arm = ArmInfo {
@@ -387,10 +411,21 @@ fn try_build_switch(arms: &[ArmInfo]) -> Option<Decision> {
         .into_iter()
         .map(|label| {
             let (needs_stdlib, inner_arms) = branch_map.remove(&label).unwrap();
+            // For type switches: a guarded arm's failure path must fall through to
+            // the catchall arms so the match stays exhaustive in Go output.
+            let decision = if matches!(kind, SwitchKind::TypeSwitch)
+                && inner_arms.iter().any(|a| a.has_guard)
+            {
+                let mut arms_with_fallback = inner_arms;
+                arms_with_fallback.extend(fallback_arms.iter().cloned());
+                build_tree(arms_with_fallback)
+            } else {
+                build_tree(inner_arms)
+            };
             SwitchBranch {
                 case_label: label,
                 needs_stdlib,
-                decision: build_tree(inner_arms),
+                decision,
             }
         })
         .collect();
@@ -463,13 +498,14 @@ fn build_chain(arms: Vec<ArmInfo>) -> Decision {
 
 /// Recursively walk a pattern, collecting checks and bindings.
 ///
-/// This replaces `traverse_pattern()` + the three visitors. Instead of
-/// running multiple passes, we collect everything in one walk.
+/// `path_ty` is the expected type of the value at `path` — used to detect
+/// when a struct pattern is matched against a Go interface (type switch).
 fn collect_checks_and_bindings(
     emitter: &mut Emitter,
     path: &AccessPath,
     pattern: &Pattern,
     typed: Option<&TypedPattern>,
+    path_ty: Option<&Type>,
     collector: &mut PatternCollector,
 ) {
     match pattern {
@@ -496,7 +532,7 @@ fn collect_checks_and_bindings(
         }
 
         Pattern::Struct { .. } => {
-            collect_struct_checks(emitter, path, pattern, typed, collector);
+            collect_struct_checks(emitter, path, pattern, typed, path_ty, collector);
         }
 
         Pattern::Tuple { elements, .. } => {
@@ -513,6 +549,7 @@ fn collect_checks_and_bindings(
                     &field_path,
                     element,
                     typed_elements.get(i).copied().flatten(),
+                    None,
                     collector,
                 );
             }
@@ -548,6 +585,7 @@ fn collect_checks_and_bindings(
                     &elem_path,
                     elem,
                     typed_prefix.get(i).copied().flatten(),
+                    None,
                     collector,
                 );
             }
@@ -564,7 +602,7 @@ fn collect_checks_and_bindings(
         }
 
         Pattern::Or { patterns, .. } => {
-            collect_or_pattern_checks(emitter, path, patterns, typed, pattern, collector);
+            collect_or_pattern_checks(emitter, path, patterns, typed, pattern, path_ty, collector);
         }
     }
 }
@@ -577,6 +615,7 @@ fn collect_or_pattern_checks(
     patterns: &[Pattern],
     typed: Option<&TypedPattern>,
     pattern: &Pattern,
+    path_ty: Option<&Type>,
     collector: &mut PatternCollector,
 ) {
     let has_bindings = Emitter::pattern_has_bindings(pattern);
@@ -592,7 +631,7 @@ fn collect_or_pattern_checks(
             .map(|(i, p)| {
                 let mut alt_collector = PatternCollector::new();
                 let tc = typed_alternatives.get(i).copied().flatten();
-                collect_checks_and_bindings(emitter, path, p, tc, &mut alt_collector);
+                collect_checks_and_bindings(emitter, path, p, tc, path_ty, &mut alt_collector);
                 alt_collector.checks
             })
             .collect();
@@ -741,6 +780,7 @@ fn collect_newtype_checks(
             &field_path,
             field,
             variant.typed_children.first().copied().flatten(),
+            None,
             collector,
         );
     }
@@ -761,6 +801,7 @@ fn collect_tuple_struct_checks(
             &field_path,
             field,
             typed_children.get(i).copied().flatten(),
+            None,
             collector,
         );
     }
@@ -835,6 +876,7 @@ fn collect_tagged_enum_checks(
                 &field_path,
                 field,
                 variant.typed_children.get(i).copied().flatten(),
+                None,
                 collector,
             );
         }
@@ -874,6 +916,7 @@ fn collect_struct_checks(
     path: &AccessPath,
     pattern: &Pattern,
     typed: Option<&TypedPattern>,
+    path_ty: Option<&Type>,
     collector: &mut PatternCollector,
 ) {
     let Pattern::Struct {
@@ -886,20 +929,30 @@ fn collect_struct_checks(
         return;
     };
 
-    let enum_info = detect_enum_info(emitter, ty, identifier, typed);
-
-    if enum_info.is_some() {
-        let alias = emitter.module_alias_for_type(ty);
-        let resolved = go_name::variant(identifier, ty, emitter.current_module(), alias.as_deref());
-        if resolved.needs_stdlib {
-            emitter.flags.needs_stdlib = true;
-        }
-        collector.checks.push(Check::EnumTag {
+    let enum_info = if path_ty.is_some_and(|st| emitter.as_interface(st).is_some()) {
+        let go_type = emitter.go_type_as_string(ty);
+        collector.checks.push(Check::TypeAssert {
             path: path.clone(),
-            tag_constant: resolved.name.clone(),
-            needs_stdlib: resolved.needs_stdlib,
+            go_type,
         });
-    }
+        None
+    } else {
+        let enum_info = detect_enum_info(emitter, ty, identifier, typed);
+        if enum_info.is_some() {
+            let alias = emitter.module_alias_for_type(ty);
+            let resolved =
+                go_name::variant(identifier, ty, emitter.current_module(), alias.as_deref());
+            if resolved.needs_stdlib {
+                emitter.flags.needs_stdlib = true;
+            }
+            collector.checks.push(Check::EnumTag {
+                path: path.clone(),
+                tag_constant: resolved.name.clone(),
+                needs_stdlib: resolved.needs_stdlib,
+            });
+        }
+        enum_info
+    };
 
     let typed_fields_map: Option<Vec<(&str, Option<&TypedPattern>)>> = match typed {
         Some(TypedPattern::Struct { pattern_fields, .. })
@@ -933,7 +986,14 @@ fn collect_struct_checks(
             enum_info.as_ref(),
             typed_variant_fields,
         );
-        collect_checks_and_bindings(emitter, &field_path, &field.value, typed_child, collector);
+        collect_checks_and_bindings(
+            emitter,
+            &field_path,
+            &field.value,
+            typed_child,
+            None,
+            collector,
+        );
     }
 }
 
@@ -986,6 +1046,7 @@ pub(super) struct ExpandedArm<'a> {
 pub(super) fn compile_expanded_arms<'a>(
     emitter: &mut Emitter,
     expanded: &'a [ExpandedArm<'a>],
+    subject_ty: &Type,
 ) -> Decision {
     let mut arm_infos: Vec<ArmInfo> = expanded
         .iter()
@@ -996,6 +1057,7 @@ pub(super) fn compile_expanded_arms<'a>(
                 &AccessPath::root(),
                 ea.pattern,
                 ea.typed_pattern,
+                Some(subject_ty),
                 &mut collector,
             );
             ArmInfo {
@@ -1045,7 +1107,14 @@ pub(crate) fn collect_pattern_info(
     typed: Option<&TypedPattern>,
 ) -> (Vec<Check>, Vec<PatternBinding>) {
     let mut collector = PatternCollector::new();
-    collect_checks_and_bindings(emitter, &AccessPath::root(), pattern, typed, &mut collector);
+    collect_checks_and_bindings(
+        emitter,
+        &AccessPath::root(),
+        pattern,
+        typed,
+        None,
+        &mut collector,
+    );
     (collector.checks, collector.bindings)
 }
 

--- a/crates/emit/src/go/patterns/matching.rs
+++ b/crates/emit/src/go/patterns/matching.rs
@@ -18,13 +18,14 @@ impl Emitter<'_> {
             self.emit_statement(output, subject);
             return;
         }
+        let subject_ty = subject.get_type();
         let subject_var = self.emit_match_subject_var(output, subject, arms);
         let guard = if matches!(subject, Expression::Literal { .. }) {
             None
         } else {
             Some(DiscardGuard::new(output, &subject_var))
         };
-        let tree_emitter = TreeEmitter::new(self, arms, ty, subject_var);
+        let tree_emitter = TreeEmitter::new(self, arms, ty, subject_var, subject_ty);
         tree_emitter.emit(output);
         if let Some(guard) = guard {
             guard.finish(output);

--- a/crates/emit/src/go/patterns/tree_emitter.rs
+++ b/crates/emit/src/go/patterns/tree_emitter.rs
@@ -4,8 +4,8 @@ use syntax::types::Type;
 use crate::Emitter;
 use crate::go::control_flow::branching::wrap_if_struct_literal;
 use crate::go::patterns::decision_tree::{
-    ChainTest, Decision, PatternBinding, SwitchKind, compile_expanded_arms, emit_tree_bindings,
-    expand_or_patterns, render_condition,
+    ChainTest, Decision, PatternBinding, SwitchBranch, SwitchKind, compile_expanded_arms,
+    emit_tree_bindings, expand_or_patterns, render_condition,
 };
 use crate::go::types::emitter::Position;
 use crate::go::utils::{inline_trivial_bindings, output_ends_with_diverge};
@@ -23,6 +23,7 @@ pub(crate) struct TreeEmitter<'a, 'e> {
     arms: &'a [MatchArm],
     ty: &'a Type,
     subject_var: String,
+    subject_ty: Type,
 }
 
 impl<'a, 'e> TreeEmitter<'a, 'e> {
@@ -31,19 +32,21 @@ impl<'a, 'e> TreeEmitter<'a, 'e> {
         arms: &'a [MatchArm],
         ty: &'a Type,
         subject_var: String,
+        subject_ty: Type,
     ) -> Self {
         Self {
             emitter,
             arms,
             ty,
             subject_var,
+            subject_ty,
         }
     }
 
     pub(crate) fn emit(mut self, output: &mut String) {
         let pre_len = output.len();
         let expanded = expand_or_patterns(self.arms);
-        let tree = compile_expanded_arms(self.emitter, &expanded);
+        let tree = compile_expanded_arms(self.emitter, &expanded, &self.subject_ty);
 
         let arm_position = self.emitter.compute_arm_position(Some(output), self.ty);
         let position = arm_position.position.clone();
@@ -89,6 +92,10 @@ impl<'a, 'e> TreeEmitter<'a, 'e> {
             unreachable!("emit_switch called on non-Switch");
         };
 
+        if matches!(kind, SwitchKind::TypeSwitch) {
+            return self.emit_type_switch(output, tree, arm_position);
+        }
+
         let subject_var = self.subject_var.clone();
 
         let switch_expression = match kind {
@@ -100,6 +107,7 @@ impl<'a, 'e> TreeEmitter<'a, 'e> {
                 let base = path.render(&subject_var);
                 wrap_if_struct_literal(base)
             }
+            SwitchKind::TypeSwitch => unreachable!("handled above"),
         };
 
         if branches.len() == 2
@@ -196,10 +204,42 @@ impl<'a, 'e> TreeEmitter<'a, 'e> {
         }
 
         write_line!(output, "switch {} {{", switch_expression);
+        self.emit_switch_body(output, branches, fallback, arm_position, &subject_var, true);
+    }
 
-        // When there's no wildcard fallback and we have branches, emit the last
-        // branch as `default:` instead of `case X:`. This satisfies Go's
-        // "all paths return" requirement without needing `panic("unreachable")`.
+    /// Emit a Go type switch: `switch x := x.(type) { case T: ... default: ... }`.
+    fn emit_type_switch(&mut self, output: &mut String, tree: &Decision, arm_position: &Position) {
+        let Decision::Switch {
+            path,
+            branches,
+            fallback,
+            ..
+        } = tree
+        else {
+            unreachable!("emit_type_switch called on non-Switch");
+        };
+
+        let subject_var = self.subject_var.clone();
+        let base = path.render(&subject_var);
+
+        write_line!(output, "switch {} := {}.(type) {{", base, base);
+        self.emit_switch_body(output, branches, fallback, arm_position, &base, false);
+    }
+
+    /// Emit case branches, the default block, the closing brace, and the
+    /// unreachable guard for a switch that has already emitted its header line.
+    ///
+    /// `track_stdlib`: when true, propagates `needs_stdlib` from branch labels
+    /// (required for enum-tag switches; not needed for type switches).
+    fn emit_switch_body(
+        &mut self,
+        output: &mut String,
+        branches: &[SwitchBranch],
+        fallback: &Option<Box<Decision>>,
+        arm_position: &Position,
+        subject_var: &str,
+        track_stdlib: bool,
+    ) {
         let use_last_as_default = fallback.is_none() && !branches.is_empty();
         let regular_branches = if use_last_as_default {
             &branches[..branches.len() - 1]
@@ -208,20 +248,18 @@ impl<'a, 'e> TreeEmitter<'a, 'e> {
         };
 
         for branch in regular_branches {
-            if branch.needs_stdlib {
+            if track_stdlib && branch.needs_stdlib {
                 self.emitter.flags.needs_stdlib = true;
             }
             write_line!(output, "case {}:", branch.case_label);
             self.emitter.enter_scope();
-
-            self.emit_decision_in_case(output, &branch.decision, arm_position, &subject_var);
-
+            self.emit_decision_in_case(output, &branch.decision, arm_position, subject_var);
             self.emitter.exit_scope();
         }
 
         let default_decision = if use_last_as_default {
             let last = branches.last().unwrap();
-            if last.needs_stdlib {
+            if track_stdlib && last.needs_stdlib {
                 self.emitter.flags.needs_stdlib = true;
             }
             Some(&last.decision)
@@ -231,7 +269,7 @@ impl<'a, 'e> TreeEmitter<'a, 'e> {
         if let Some(decision) = default_decision {
             let pre = output.len();
             self.emitter.enter_scope();
-            self.emit_decision_in_case(output, decision, arm_position, &subject_var);
+            self.emit_decision_in_case(output, decision, arm_position, subject_var);
             self.emitter.exit_scope();
             if output.len() > pre {
                 output.insert_str(pre, "default:\n");

--- a/tests/spec/emit/patterns.rs
+++ b/tests/spec/emit/patterns.rs
@@ -1,4 +1,5 @@
 use crate::assert_emit_snapshot;
+use crate::assert_emit_snapshot_with_go_typedefs;
 
 #[test]
 fn tuple_struct_pattern_in_match() {
@@ -562,4 +563,68 @@ fn main() {
 }
 "#;
     assert_emit_snapshot!(input);
+}
+
+#[test]
+fn match_on_go_interface_emits_type_switch() {
+    let input = r#"
+import "go:example.com/events"
+
+fn handle(e: events.Event) -> int {
+  match e {
+    events.Click { x, y } => x + y,
+    events.KeyPress { key } => key.length(),
+    _ => 0,
+  }
+}
+"#;
+    let typedef = r#"
+pub interface Event {}
+pub struct Click { pub x: int, pub y: int }
+pub struct KeyPress { pub key: string }
+"#;
+    assert_emit_snapshot_with_go_typedefs!(input, &[("go:example.com/events", typedef)]);
+}
+
+#[test]
+fn match_on_go_interface_guarded_arm() {
+    let input = r#"
+import "go:example.com/events"
+
+fn handle(e: events.Event, active: bool) -> int {
+  match e {
+    events.Click { x, y } => x + y,
+    events.KeyPress { .. } if active => -1,
+    _ => 0,
+  }
+}
+"#;
+    let typedef = r#"
+pub interface Event {}
+pub struct Click { pub x: int, pub y: int }
+pub struct KeyPress { pub key: string }
+"#;
+    assert_emit_snapshot_with_go_typedefs!(input, &[("go:example.com/events", typedef)]);
+}
+#[test]
+fn match_on_go_interface_wildcard_only_arm() {
+    let input = r#"
+import "go:example.com/events"
+
+fn handle(e: events.Event) -> int {
+  match e {
+    events.Click { x, y } => x + y,
+    events.KeyPress { .. } => -1,
+    events.Resize { .. } => 0,
+    _ => 0,
+  }
+}
+"#;
+    let typedef = r#"
+pub interface Event {}
+pub struct Click { pub x: int, pub y: int }
+pub struct KeyPress { pub key: string }
+pub struct Resize { pub width: int, pub height: int }
+"#;
+    assert_emit_snapshot_with_go_typedefs!(input, &[("go:example.com/events", typedef)]);
 }

